src/hotspot/share/memory/allocation.hpp

```diff
@@ -431,13 +430,13 @@ class ResourceObj ALLOCATION_SUPER_CLASS_SPEC {
 
   void* operator new(size_t size) throw() {
       address res = (address)resource_allocate_bytes(size);
-      DEBUG_ONLY(set_allocation_type(res, RESOURCE_AREA);)
+      DEBUG_ONLY(_thread_last_allocated = RESOURCE_AREA;)
       return res;
   }
 
   void* operator new(size_t size, const std::nothrow_t& nothrow_constant) throw() {
       address res = (address)resource_allocate_bytes(size, AllocFailStrategy::RETURN_NULL);
-      DEBUG_ONLY(if (res != NULL) set_allocation_type(res, RESOURCE_AREA);)
+      DEBUG_ONLY(if (res != NULL) _thread_last_allocated = RESOURCE_AREA;)
```

Comment:
Maybe we should also guard against the possibility of nested allocations, which may trash `_thread_last_allocated`?

```
#define PUSH_RESOURCE_OBJ_ALLOC_TYPE(t) \
  assert(_thread_last_allocated == STACK_OR_EMBEDDED, "must not be nested"); \
  DEBUG_ONLY(_thread_last_allocated = t); \

...
  if (res != NULL) {
    PUSH_RESOURCE_OBJ_ALLOC_TYPE(RESOURCE_AREA);
  }
```

Similarly, the `ResourceObj` constructor should use a corresponding `POP_RESOURCE_OBJ_ALLOC_TYPE` macro.

Reply:
I added a `set_type` method that ensures that the `_thread_last_allocated` always transition over a `STACK_OR_EMBEDDED`. I did *not* create a PUSH/POP macro pair because i believe it would give the false impression that we are doing a stack operation. 

Other than that I also made `allocation_type` use a `u8` as backing type. I also removed the now unused `allocation_mask` and the now unimportant detail that `STACK_OR_EMBEDDED = 0`.

Reply:
Hmm, u8 was not what I was thinking, I will change that to a uint8_t in the next update...

Reply:
I hit the new assert when not on Linux, I guess it has to do with the initialization of the thread local variable.

Reply:
Thanks Ioi for making me adding the assert!!! The sequencing of the allocation function and the arguments to the constructor is not what I thought, so my "solution" is not working. I am unsure how to resolve this in a good way. We could probably have a small thread local collection of (type, address) pairs, but I wonder if it is not better removing this debug information altogether. It is to my knowledge only used in GrowableArray (to limit the type of its internal allocations) and to hinder delete on resource allocated objects.

Reply:
A collection of (type, address) pairs is still problematic.  It still requires assuming that the address of the derived object is the same as the address of the ResourceObj subobject, which isn't guaranteed, though the current mechanism also depends on that being true.  I think there might be other places in HotSpot where we're making that assumption too, unfortunately.